File: minWindow.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function (s, t) {
  if (s.length < t.length || (s !== t && s.length === 1 && t.length === 1))
    return "";
  if (s === t) return s;

  let [left, right, requiredChars, foundedChars, resLen, res] = [
    0,
    0,
    0,
    0,
    Number.MAX_SAFE_INTEGER,
    [0, 0],
  ];

  const [tCount, windowChars] = [new Map(), new Map()];

  for (const char of t) {
    tCount.set(char, (tCount.get(char) || 0) + 1);
  }
  requiredChars = tCount.size;

  for (; right < s.length; right++) {
    const char = s[right];

    tCount.has(char) && windowChars.set(char, (windowChars.get(char) || 0) + 1);

    if (tCount.has(char) && windowChars.get(char) === tCount.get(char)) {
      foundedChars += 1;
    }

    // Find an answer
    while (foundedChars === requiredChars) {
      if (right - left + 1 < resLen) {
        resLen = right - left + 1;
        res = [left, right];
      }

      windowChars.set(s[left], (windowChars.get(s[left]) || 0) - 1);

      if (
        tCount.has(s[left]) &&
        windowChars.get(s[left]) < tCount.get(s[left])
      ) {
        foundedChars -= 1;
      }
      left++;
    }
  }
  return resLen === Number.MAX_SAFE_INTEGER
    ? ""
    : s.substring(res[0], res[1] + 1);
};

console.log(minWindow("abc", "ac") === "abc");
console.log(minWindow("acbbaca", "aba") === "baca");

File: JavaWorkSpace/workspace/src/JavaPlayGround/MinWindow.java
package JavaPlayGround;

import java.util.*;

public class MinWindow {
    public String minWindow(String s, String t) {
        if (s.equals(t))
            return s;

        int[] minWin = { 0, Integer.MAX_VALUE - 1 };
        Map<Character, Integer> tCount = new HashMap<>();
        Map<Character, Integer> windowCount = new HashMap<>();

        for (char c : t.toCharArray()) {
            tCount.put(c, tCount.getOrDefault(c, 0) + 1);
        }

        int requiredChars = tCount.size();
        int windowChars = 0;

        int left = 0;
        int right = 0;

        while (right < s.length()) {
            char charRight = s.charAt(right);

            if (tCount.containsKey(charRight)) {
                windowCount.put(charRight, windowCount.getOrDefault(charRight, 0) + 1);
                if (windowCount.get(charRight).equals(tCount.get(charRight))) {
                    windowChars++;
                }
            }

            while (windowChars >= requiredChars) {
                if (right - left + 1 < minWin[1] - minWin[0] + 1) {
                    minWin[0] = left;
                    minWin[1] = right;
                }

                char charLeft = s.charAt(left);

                if (windowCount.containsKey(charLeft)) {
                    windowCount.put(charLeft, windowCount.get(charLeft) - 1);
                    if (windowCount.get(charLeft) < tCount.get(charLeft)) {
                        windowChars--;
                    }
                }

                left++;
            }

            right++;
        }

        return minWin[1] == Integer.MAX_VALUE - 1
                ? ""
                : s.substring(minWin[0], minWin[1] + 1);
    }

    public static void main(String[] args) {
        System.out.println(new MinWindow().minWindow("ADOBECODEBANC", "ABC"));
        System.out.println(new MinWindow().minWindow("aa", "aa"));

    }
}
